Create documentation for filesystem repositories (#4287)
alcionai/corso: website/docs/setup/repos.md

Within a repository, Corso uses
AES256-GCM-HMAC-SHA256 to encrypt data at rest using keys that are derived from the repository passphrase.
Data in flight to and from the repository is encrypted via TLS.

> Repositories are supported on the following object storage systems:

Reviewer: ```suggestion
Repositories are supported on the following storage backends:
```

Corso also supports the use of object storage systems with no TLS certificate or with self-signed
TLS certificates with the `--disable-tls` or `--disable-tls-verification` flags.
[These flags](../../cli/corso-repo-init-s3) should only be used for testing.

## Filesystem Storage

> Corso supports creating a repository on local or network attached filesystem.

Reviewer 2: ```suggestion
Corso supports creating a repository on a local or network attached filesystem.
```

:::note

> Filesystem repositories aren't S3 compatible.

Reviewer 2: For my own edification, what does this mean?

Author: It means we cannot seamlessly transfer repository data between different environments, for e.g. from an on-prem server to a cloud provider, without doing some complex data transformation.

Minio on the other hand is [S3 compatible.](https://min.io/product/s3-compatibility)

Reviewer 3: You should expand on what this means in the note.

Adding something along the lines of "Copying a local repository directly to AWS S3 or the other way around is not supported and will not work."

Not that we are doing it, but we probably should not be encouraging direct repo operations for Minio or other S3 compatible storage either. It should work but is asking for trouble.

:::

> ### Filesystem storage prerequisites
> 
> None.
> 

Reviewer 2: Can this section be omitted?
```suggestion
```

Author: Fair. I added it so that it's clear to the reader that there are no prerequisites. Compared to a perceived missing section.

### Initialize filesystem repository

Before first use, you need to initialize a Corso repository with `corso repo init filesystem`. See the command details
[here](../../cli/corso-repo-init-filesystem). Corso will create the directory structure if necessary, including any
missing parent directories.

<Tabs groupId="os">
<TabItem value="win" label="Powershell">

  ```powershell
  # Initialize the Corso Repository
  $Env:CORSO_PASSPHRASE = 'CHANGE-ME-THIS-IS-INSECURE'
  .\corso repo init filesystem --path C:\Users\user\corso-test
  ```

</TabItem>
<TabItem value="unix" label="Linux/macOS">

  ```bash
  # Initialize the Corso Repository
  export CORSO_PASSPHRASE="CHANGE-ME-THIS-IS-INSECURE"

>   ./corso repo init filesystem --path $HOME/corso-test

Reviewer 2: General preference is for commands to look more like regular use and less like PoC values.  In this case, can we replace `-test` with some other suffix?  `-backup`?
```suggestion
  ./corso repo init filesystem --path $HOME/corso-test
```

Author: yeah, thats a great point. I noticed that we use `corso-test` in a few other places - [example](https://corsobackup.io/docs/quickstart/#create-a-corso-repository), was trying to stay consistent. I'll fix this as well as other occurrences.

  ```

</TabItem>
<TabItem value="docker" label="Docker">

<CodeBlock language="bash">{
`# Initialize the Corso Repository
export CORSO_PASSPHRASE="CHANGE-ME-THIS-IS-INSECURE"
docker run --env-file $HOME/.corso/corso.env \\
  --volume $HOME/.corso:/app/corso ghcr.io/alcionai/corso:${Version()} \\
  repo init filesystem --path /path/in/container`

> }</CodeBlock>

Reviewer 3: I think this example needs work. 

With the above you will most likely endup with a local repo on the container file system which is going to be ephemeral. I think this is easy to miss for most people. 

I think you want to develop the example so that the location where a filesystem location from the host is mapped to a specific location inside the container. The simplest would be to map another volume

Author: Thanks! I've fixed this by adding a volume. Confirmed that this works. 

```
# Initialize the Corso Repository
export CORSO_PASSPHRASE="CHANGE-ME-THIS-IS-INSECURE"
docker run --env-file $HOME/.corso/corso.env \
  --volume $HOME/.corso:/app/corso ghcr.io/alcionai/corso:unreleased-280ba3ace \
  --volume /path/on/host/corso-repo:/path/in/container/corso-repo \
  repo init filesystem --path /path/in/container/corso-repo
```

</TabItem>
</Tabs>

### Connect to a filesystem repository

If a repository already exists, you can connect to it with `corso repo connect filesystem`. See the command details
[here](../../cli/corso-repo-connect-filesystem).

<Tabs groupId="os">
<TabItem value="win" label="Powershell">

  ```powershell
  # Connect to the Corso Repository

>   .\corso repo connect filesystem --path C:\Users\user\corso-test

Reviewer 2: Is path required to connect?

Author: yep

  ```

</TabItem>
<TabItem value="unix" label="Linux/macOS">

  ```bash
  # Connect to the Corso Repository
  ./corso repo connect filesystem --path $HOME/corso-test
  ```

</TabItem>
<TabItem value="docker" label="Docker">

<CodeBlock language="bash">{
`# Connect to the Corso Repository
docker run --env-file $HOME/.corso/corso.env \\
  --volume $HOME/.corso:/app/corso ghcr.io/alcionai/corso:${Version()} \\
  repo connect filesystem --path /path/in/container`

> }</CodeBlock>

Reviewer 3: Same as above. All docker examples need to take into account the mapping of the external location.

</TabItem>
</Tabs>

### Remarks

* Filesystem repositories are quick to setup and can be suitable for smaller scale projects or
testing environments. However, they don't offer the same level of interoperability as a
S3 compatible object storage.

> * Repository `--prefix` options aren't supported.

Reviewer 2: Could we manufacture this support one day?  Seems like it's equivalent to hacking in a subdirectory beneath the provided location.

Author: It's not supported by kopia.

Reviewer 2: I thought it was a S3 construct?
